core/src/main/scala/api/sfs.scala

```diff
@@ -25,15 +30,19 @@ trait Filesystem {
    *  things too eagerly, nor caching data for too long.
    */
   def resolve(path: Path): Key
-  def metadata(key: Key): Metadata
-  def lookup(key: Key): Data
+  def metadata(key: Key): M[Metadata]
+  def lookup(key: Key): M[Data]
```

Comment:
Thinking about the 'single method' or 'everything through metadata' approach will probably make things more complicated. Imagine a filesystem that has it's metadata stored on a remote server, it would then give you the following:

```
def resolve(path: Path): Metadata
...
metadata[Future[Option[Size]]]
metadata[Future[Option[Type]]
```

Where `Future` is the effect of stuff being 'not fast enough' and `Option` the effect of the metadata not available (when only know in future). If you think about it, their signatures are `Option[Future[Option[X]]]`. 

```
def resolve(path: Path): Future[Metadata] // type M[A] = Future[A]
...
metadata[Size]
metadata[Type]
```

Now their signatures are effectively `Future[Option[X]]`. On top of that, even if different file systems have different effects, their metadata types are more likely to line up.